src/Compilers/CSharp/Test/Semantic/Semantics/StaticNullChecking.cs

```diff
@@ -16829,33 +16809,85 @@ class A : System.Attribute
         }
 
         [Fact]
-        public void UnassignedParameterField()
+        public void UnassignedOutParameterClass()
         {
             var source =
 @"class C
 {
-    static void F(out S s)
+    static void G(out C? c)
     {
-        C c;
-        c = s.F;
-        s.F.ToString();
+        c.ToString(); // 1
+        c = null;
+        c.ToString(); // 2
+        c = new C();
+        c.ToString(); // 3
     }
-}
-struct S
+}";
+            var comp = CreateStandardCompilation(source, parseOptions: TestOptions.Regular8);
+            comp.VerifyDiagnostics(
+                // (5,9): error CS0269: Use of unassigned out parameter 'c'
+                //         c.ToString(); // 1
+                Diagnostic(ErrorCode.ERR_UseDefViolationOut, "c").WithArguments("c").WithLocation(5, 9),
+                // (7,9): warning CS8602: Possible dereference of a null reference.
+                //         c.ToString(); // 2
+                Diagnostic(ErrorCode.WRN_NullReferenceReceiver, "c").WithLocation(7, 9),
+                // (9,9): warning CS8602: Possible dereference of a null reference.
+                //         c.ToString(); // 3
+                Diagnostic(ErrorCode.WRN_NullReferenceReceiver, "c").WithLocation(9, 9));
```

Comment:
>Diagnostic(ErrorCode.WRN_NullReferenceReceiver, "c").WithLocation(9, 9)); [](start = 16, length = 73)

This warning looks unexpected. Was ```c``` assigned on the previous line ```c = new C();```?

Reply:
Good catch. It looks like the state for `c` is being set to `false` in `TrackNullableStateForAssignment` since the parameter is passed by ref. Any idea what the comment means?
```
    this.State[slot] = isByRefTarget ?
        // Since reference can point to the heap, we cannot assume the value is not null after this assignment,
        // regardless of what value is being assigned. 
        (targetIsNullable == true) ? (bool?)false : null :
        valueIsNotNull;
```